src/Integration/Connection/ConnectionController.cs

```diff
@@ -0,0 +1,202 @@
+            Debug.Assert(connectionToRefresh != null, "Expecting either to be connected to get a connection to connect to");
+
+            // Any existing connection will be disconnected and disposed, so create a copy and use it to connect
+            this.EstablishConnection(connectionToRefresh.Clone());
+        }
+        #endregion
+
+        #region Don't warn again command
+        private void ShowNuGetWarning(ProgressControllerResult executionResult)
+        {
+            if (executionResult == ProgressControllerResult.Succeeded && this.settings.ShowServerNuGetTrustWarning)
+            {
+                this.host.ActiveSection?.UserNotifications?.ShowNotificationWarning(Strings.ServerNuGetTrustWarningMessage, NotificationIds.WarnServerTrustId, this.DontWarnAgainCommand);
+            }
+        }
+
+        private void OnDontWarnAgain()
+        {
+            this.settings.ShowServerNuGetTrustWarning = false;
+            this.host.ActiveSection?.UserNotifications?.HideNotification(NotificationIds.WarnServerTrustId);
+        }
+
+        private bool OnDontWarnAgainStatus()
+        {
+            return this.settings != null;
+        }
+        #endregion
+
+        #region IConnectionInformationProvider
+
+        ConnectionInformation IConnectionInformationProvider.GetConnectionInformation(ConnectionInformation currentConnection)
+        {
+            var dialog = new ConnectionInformationDialog();
+            return dialog.ShowDialog(currentConnection);
+        }
+
+        #endregion
+
+        #region IConnectionWorkflowExecutor
+        private void EstablishConnection(ConnectionInformation connectionInfo)
+        {
+            Debug.Assert(connectionInfo != null);
+
+            this.LastAttemptedConnection = connectionInfo;
+
+            this.WorkflowExecutor.EstablishConnection(connectionInfo, this.host.VisualStateManager.SetProjects);
+        }
+
+        void IConnectionWorkflowExecutor.EstablishConnection(ConnectionInformation information, ConnectedProjectsCallback connectedProjectsChanged)
+        {
+            ConnectionWorkflow workflow = new ConnectionWorkflow(this.host, this.ConnectCommand, connectedProjectsChanged);
+            IProgressEvents progressEvents = workflow.Run(information);
+            this.SetConnectionInProgress(progressEvents);
+        }
+
+        internal /*for testing purposes*/ void SetConnectionInProgress(IProgressEvents progressEvents)
+        {
+            this.IsConnectionInProgress = true;
+
+            ProgressNotificationListener progressListener = new ProgressNotificationListener(this.ServiceProvider, progressEvents);
+            progressListener.MessageFormat = Strings.ConnectingToSonarQubePrefixMessageFormat;
+
+            progressEvents.RunOnFinished(result =>
+            {
+                progressListener.Dispose();
+                this.IsConnectionInProgress = false;
+                this.ShowNuGetWarning(result);
```

Comment:
Another issue that i was able to address (previously was defined in the workflow when in fact it should be defined on the controller).